### PR TITLE
[mlir][vector] Propagate scalability to gather/scatter ptrs vector

### PR DESCRIPTION
**Pull request by c-rhodes**
In convert-vector-to-llvm the first operand (vector of pointers holding
all memory addresses to read) to the masked.gather (and scatter)
intrinsic has a fixed vector type.

This may result in intrinsics where the scalable flag has been dropped:
```
  %0 = llvm.intr.masked.gather %1, %2, %3 {alignment = 4 : i32}
    : (!llvm.vec<4 x ptr>, vector<[4]xi1>, vector<[4]xi32>) -> vector<[4]xi32>
```
Fortunately the operand is overloaded on the result type so we end up
with the correct IR when lowering to LLVM, but this is still incorrect.
This patch fixes it by propagating scalability.

### CODE REVIEW

**Review by banach-space — APPROVED** [review]
Makes sense, thanks for the fix! LGTM

**Review by dcaballe — APPROVED** [review]
Thanks! Is there a way to check this in the verifier?

**Comment by c-rhodes** [comment]
> Thanks! Is there a way to check this in the verifier?

I've added verifiers for gather/scatter intrinsic to check this. Generally the intrinsics are quite loose with constraints, seems like a bit of a rabbit hole.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp [hunk 1]
```diff
@@ -3076,6 +3076,36 @@ void InlineAsmOp::getEffects(
   }
 }
 
+//===----------------------------------------------------------------------===//
+// masked_gather (intrinsic)
+//===----------------------------------------------------------------------===//
+
+LogicalResult LLVM::masked_gather::verify() {
+  auto ptrsVectorType = getPtrs().getType();
+  Type expectedPtrsVectorType =
```
Comment (dcaballe) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:3085]: nit: a comment about the specific issue we want to verify here would help understanding

**Review by dcaballe — APPROVED** [review]
Awesome, thanks!

File: mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp [hunk 2]
```diff
@@ -102,11 +102,13 @@ static LogicalResult isMemRefTypeSupported(MemRefType memRefType,
 static Value getIndexedPtrs(ConversionPatternRewriter &rewriter, Location loc,
                             const LLVMTypeConverter &typeConverter,
                             MemRefType memRefType, Value llvmMemref, Value base,
-                            Value index, uint64_t vLen) {
+                            Value index, VectorType vectorType) {
   assert(succeeded(isMemRefTypeSupported(memRefType, typeConverter)) &&
          "unsupported memref type");
   auto pType = MemRefDescriptor(llvmMemref).getElementPtrType();
-  auto ptrsType = LLVM::getFixedVectorType(pType, vLen);
+  auto ptrsType =
```
Comment (MacDue) [inline mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:109]: Assert vType is 1-D?

**Review by MacDue — APPROVED** [review]

**Pull request merged**